### ISSUE TITLE
a few suggestions for the architectural upgrade

### ISSUE THREAD

**Issue by danelee2601** [jishengpeng/WavTokenizer#74]
Hi, thanks for the cool paper and clean implementation.
After going through them, I got some thoughts and wanted to share with you.

1. Use better and scalable quantizer: Binary quantizers like the one in this paper, https://arxiv.org/abs/2409.16211, have been shown super scalable and superior to the traditional quantizers with near 100% codebook utilization even at the codebook size of 2**18. Highly recommend this.
2. The encoder practically has dilation size of 1 since the number of residual blocks is set to 1. Probably increasing this should help with the wider receptive field size. Although the LSTM produces the wide receptive field, I think the larger dilation should help. In my hindsight, maybe because the downsampling rate is so high that it already results in a large receptive field, so dilation=1 is just fine..

**Comment by danelee2601** [comment]
<img width="1103" alt="Image" src="https://github.com/user-attachments/assets/95410937-ea68-4804-a0b2-7b26af33c2ad" />

I used the codebook size of 2**14 for the BLFQ.

**Comment by danelee2601** [comment]
I also noticed that your code contains the loss weight decay on the mel reconstruction loss, which I think is a very reasonable thing to do to some degree. But I think there seems to be a bug in the code. The current decay function results in decay and reverts it back after the half of the training steps. See the plot.

![Image](https://github.com/user-attachments/assets/f8c3088e-9376-4e3b-8fcc-51197a8cb973)


You probably wanna do something like this:

![Image](https://github.com/user-attachments/assets/9188351a-0bfd-4816-8497-521a0f80ea25)

```python
import math
trainer_max_steps = 2000000
num_warmup_steps = 0

def mel_loss_coeff_decay_original(current_step, num_cycles=0.5):
    max_steps = trainer_max_steps // 2
    if current_step < num_warmup_steps:
        return 1.0
    progress = float(current_step - num_warmup_steps) / float(max(1, max_steps - num_warmup_steps))
    return max(0.0, 0.5 * (1.0 + math.cos(2.0*math.pi * float(num_cycles) * progress)))

def mel_loss_coeff_decay_fixed(current_step, num_cycles=0.5):
    max_steps = trainer_max_steps // 2
    if current_step < num_warmup_steps:
        return 1.0
    if current_step > max_steps:
        current_step = max_steps
    progress = float(current_step - num_warmup_steps) / float(max(1, max_steps - num_warmup_steps))
    return max(0.0, 0.5 * (1.0 + math.cos(2.0*math.pi * float(num_cycles) * progress)))
```

**Comment by danelee2601** [comment]
Additionally, if you have experimented with the loss coefficient decay already, can you share your experience in terms of the result?

**Comment by jishengpeng** [comment]
> Additionally, if you have experimented with the loss coefficient decay already, can you share your experience in terms of the result?

Thank you very much for your attention and experiments. 

1. Regarding the first point, expanding the codebook space and adopting more advanced vector quantization strategies—such as SimVQ and the method you mentioned—can indeed further improve performance on top of the wavtokenizer framework. However, since we aim to balance reconstruction and generation tasks, we advise against excessively enlarging the codebook space.

2. We have not tuned the dilation parameter thus far, so we sincerely appreciate your suggestion.

3. The use of loss weight decay on the mel reconstruction loss is indeed a critical aspect. In our view, both the current design and your proposed decay approach are reasonable. Furthermore, we can share that in one of our stable model versions, significantly increasing the mel_loss weight in the configuration file—for example, from 45 to 900—can lead to notable improvements in reconstruction quality.

**Comment by danelee2601** [comment]
Thanks for sharing your opinions! Really appreciated.
